### PR TITLE
Add handleRedirectCallback

### PR DESCRIPTION
**Pull request by frederikprijck**
When using Auth0-SPA-JS, you need to call handleRedirectCallback on page-load, see https://auth0.com/docs/quickstart/spa/vanillajs/interactive

Basically, the code that was missing was:

```js
  if (location.search.includes("state=") && 
      (location.search.includes("code=") || 
      location.search.includes("error="))) {
    await auth0Client.handleRedirectCallback();
    window.history.replaceState({}, document.title, "/");
  }
```

With that in place, I was able to login on firefox and safari (note I used localhost:3000, but that shouldnt matter)

image
image

Also note that I am not familiar with the abstractions in this repository, so this solution is not 100% ready I would think. But it shows what can be done to make it work and you can use it to build your actual solution.

---
Also, I recommend a different approach than what we have now on page load. Currently, on page-load it calls `signIn({ silent: true })` and `handleRedirectCallback()`. I recommend doing the following:

- If `location.search` has `state` and (`code` or `error`), call `handleRedirectCallback`
- Else, call `signIn({silent: true})`

That would make more sense imho, but may require a bit rework in the actions / events you have, so I didnt bother doing that.

### CODE REVIEW

**Comment by chrisrabe** [comment]
Wow. This works. Thank you. I must've missed this part in the Auth0 documentation. 

I've modified your solution a bit so that it's included in the signIn function since we need to emit the access token on the signIn event but from my testing, it seems to be have as I wanted to. Just wanted to get your thoughts on this PR.

https://github.com/chrisrabe/distributable-auth0-sdk/pull/2/files

Also, how does this handleRedirectCallback work? Where does it store the session when that function is called?

**Comment by frederikprijck** [comment]
> Just wanted to get your thoughts on this PR.

Looks like it should work, I also considered just adding it into signIn, but figured thats for you to figure out and I just added it in the first way I tried based on the rest.

> Also, how does this handleRedirectCallback work? Where does it store the session when that function is called?

https://github.com/auth0/auth0-spa-js/blob/main/src/Auth0Client.ts#L485
handleRedirectCallback is not that different from getTokenSilently, and stores everything in the same way.

But, what exactly do you mean with `Where does it store the session when that function is called` ?

**Comment by chrisrabe** [comment]
> But, what exactly do you mean with `Where does it store the session when that function is called` ?

Sorry, let me rephrase.

Let's say that the JS bundle that's generated from this code is hosted in the domain `account-sdk.example.com/sdk.js`. A sign in button component is created for teams to use. This sign in button relies on the sdk.js file to determine whether you're authenticated or not, as well as handle sign in and sign out.

One team manages `example.com`, and another team manages `contact.example.com`. Both embeds the same sign in button and the sdk.js file.

We have a use case where if someone logs in successfully to example.com, they should also be authenticated when they navigate to the subdomain contact.example.com.

From playing around with the Auth0.js SDK, Auth0 seems to store session information inside a cookie under the tenant auth0 domain and uses that information when the access token is not available in the local storage. This caused problems for us in the use case I gave above because Safari and Firefox interprets the cookies stored under auth0 domain as a third party cookie. We resolved it by setting up custom domains because the cookie is now stored under the same domain name as the parent.

With the handleRedirectCallback, I'm just wondering if we still need those custom domains to fix the cross origin policy issues that Safari and Firefox has by checking where the session is stored (is it still in the tenant auth0 domain?)

**Comment by chrisrabe** [comment]
Ran some experiments to answer my question. I still need to use custom domains. 

handleRedirectCallback just puts the access token and idToken into the local storage using the query parameters attached to the redirect URL. This _only_ fixes the issue where the authenticated state is not shown after redirecting back using browsers that have strict cross-origin policies (e.g. Safari, Firefox, Brave and Chrome Incognito). 

It doesn't really do anything special with the user's session.

If I need to retain authenticated state between subdomains, I _have_ to use custom domains otherwise the user will have to keep logging in again and again.

**Comment by chrisrabe** [comment]
Anyway, thanks for pointing out the missing `handleRedirectCallback` @frederikprijck. This fixes a few pain points on our end.

**Comment by frederikprijck** [comment]
Hey, sorry i missed this.

Yes, custom domains is always a good idea, and needed for anything that doesnt like third party cookies.
